mas-foundation/src/main/java/com/ca/mas/core/oauth/CodeVerifierCache.java

```diff
@@ -8,44 +8,45 @@
 
 package com.ca.mas.core.oauth;
 
+import com.ca.mas.core.storage.sharedstorage.SharedPreferencesUtil;
+
 /**
  * Temporary cache to store code verifier
  */
 public class CodeVerifierCache {
 
     private static CodeVerifierCache instance = new CodeVerifierCache();
-
-    private String state;
-    private String codeVerifier;
+    private SharedPreferencesUtil prefUtil = null;
 
     private CodeVerifierCache() {
+        prefUtil = new SharedPreferencesUtil("codeverifier");
     }
 
     public static CodeVerifierCache getInstance() {
         return instance;
     }
 
     public void store(String state, String codeVerifier) {
-        this.state = state;
-        this.codeVerifier = codeVerifier;
+        prefUtil.save("state", state);
+        prefUtil.save("code", codeVerifier);
     }
 
     public String take(String state) {
-        if (this.state == null && state != null
-                || this.state != null && !this.state.equals(state)) {
+        if (prefUtil.getString("state") == null && state != null
+                || prefUtil.getString("state")!= null && !prefUtil.getString("state").equals(state)) {
             throw new IllegalStateException("OAuth State Mismatch");
         }
-        String cv = this.codeVerifier;
-        this.state = null;
-        this.codeVerifier = null;
+        String cv = prefUtil.getString("code");
```

Comment:
Having the one-one relation between state and codeverifier, you should think of retrieving the codeverifier using state.

And, what is the expectation of **take** method. Is it both retrieve and delete functionality?

Reply:
Yes, take method is used for both retrieve and delete. As code verifier will be generated while checking the AppAuth.